File: OnlyCue/LTC/LTCEncoder.swift
import AVFoundation

/// Generates LTC audio: builds the 80-bit `LTCFrame` for a `Timecode`,
/// biphase-mark-modulates it, and lays the result down as `Float` PCM at a
/// given sample rate.
///
/// The 80-bit frame is split into 160 half-bit slots; slot `k` occupies samples
/// `[round(k·R) … round((k+1)·R))` where `R = sampleRate / (160·fps)`. So the
/// timing is exact at any sample rate — integer at 48 kHz for 24 / 25 / 30 fps,
/// and the rounding distributes the remainder cleanly when it isn't (e.g.
/// 44.1 kHz, or the 12.5-samples-per-half-bit case of 24 fps at 48 kHz, which is
/// integer at the *bit* level). `fps` here is the timeline rate — 30 for both
/// 30-ND and 30-DF (ADR-019: drop-frame is a labelling convention, not 29.97).
enum LTCEncoder {

    /// Standalone/test fallback amplitude for callers that build a schedule
    /// without routing. The **product** default lives in
    /// `LTCRoutingSettings.defaultAmplitude` (0.9) and is always passed
    /// explicitly on the live output path (#651).
    static let defaultAmplitude: Float = 0.8

    /// Float PCM samples (mono, `±amplitude`) for one LTC frame at `timecode`,
    /// continuous from `startLevel`. Returns the samples and the signal level
    /// after the last sample so consecutive frames can be modulated continuously
    /// (pass it back as the next call's `startLevel`).
    static func samples(
        for timecode: Timecode,
        sampleRate: Double,
        amplitude: Float = defaultAmplitude,
        startLevel: Bool = false
    ) -> (samples: [Float], endLevel: Bool) {
        samples(
            for: LTCFrame(timecode: timecode),
            framesPerSecond: timecode.rate.framesPerSecond,
            sampleRate: sampleRate,
            amplitude: amplitude,
            startLevel: startLevel
        )
    }

    /// The same modulation, driven by an explicit 80-bit word rather than by a
    /// `Timecode`. Nothing in the app calls this — `LTCFrame(timecode:)` always
    /// produces a well-formed word — but `golden/ltc-decode-v1.json` needs it:
    /// its structural-corruption cases are built by flipping bits in a frame and
    /// re-modulating, so that "bad parity" means a genuinely mis-parity word on
    /// the wire rather than a smeared waveform. Going through the production
    /// encoder is the point; a test-local modulator would pin itself.
    static func samples(
        for frame: LTCFrame,
        framesPerSecond: Int,
        sampleRate: Double,
        amplitude: Float = defaultAmplitude,
        startLevel: Bool = false
    ) -> (samples: [Float], endLevel: Bool) {
        precondition(sampleRate > 0, "sample rate must be positive")
        let halfBitSamples = sampleRate / (160.0 * Double(framesPerSecond))
        let high = amplitude
        let low = -amplitude

        var level = startLevel
        var samples: [Float] = []
        samples.reserveCapacity(Int((sampleRate / Double(framesPerSecond)).rounded()) + 2)
        var slot = 0

        func emitSlot() {
            let start = Int((Double(slot) * halfBitSamples).rounded())
            let end = Int((Double(slot + 1) * halfBitSamples).rounded())
            samples.append(contentsOf: repeatElement(level ? high : low, count: max(0, end - start)))
            slot += 1
        }

        for bit in frame.bits {
            level.toggle()          // bit-boundary transition (always)
            emitSlot()
            if bit { level.toggle() }   // mid-bit transition for a 1
            emitSlot()
        }
        return (samples, level)
    }

    /// One LTC frame as an `AVAudioPCMBuffer` (`.pcmFormatFloat32`, mono, `sampleRate`).
    static func makeBuffer(
        for timecode: Timecode,
        sampleRate: Double,
        amplitude: Float = defaultAmplitude
    ) -> AVAudioPCMBuffer? {
        let format = AVAudioFormat(commonFormat: .pcmFormatFloat32, sampleRate: sampleRate, channels: 1, interleaved: false)
        let pcm = samples(for: timecode, sampleRate: sampleRate, amplitude: amplitude).samples
        guard let format,
              !pcm.isEmpty,
              let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: AVAudioFrameCount(pcm.count)),
              let channel = buffer.floatChannelData else { return nil }
        buffer.frameLength = AVAudioFrameCount(pcm.count)
        pcm.withUnsafeBufferPointer { source in
            if let base = source.baseAddress { channel.pointee.update(from: base, count: pcm.count) }
        }
        return buffer
    }
}

File: windows/OnlyCue.Core/Ltc/LtcEncoder.cs
namespace OnlyCue.Core.Ltc;

/// <summary>
/// Generates LTC audio: builds the 80-bit <see cref="LtcFrame"/> for a
/// <see cref="Timecode"/>, biphase-mark-modulates it, and lays the result down as
/// <c>float</c> PCM at a given sample rate.
/// </summary>
/// <remarks>
/// <para>
/// Mirrors the Swift <c>LTCEncoder.samples(for:sampleRate:amplitude:startLevel:)</c>
/// (<c>OnlyCue/LTC/LTCEncoder.swift</c>); macOS is the source of truth and
/// <c>golden/ltc-wire-v1.json</c> is the contract. The <c>AVAudioPCMBuffer</c>
/// half of the Swift type is Apple-only and has no mirror.
/// </para>
/// <para>
/// The 80-bit frame is split into 160 half-bit slots; slot <c>k</c> occupies
/// samples <c>[round(k·R) … round((k+1)·R))</c> where
/// <c>R = sampleRate / (160·fps)</c>. <c>R</c> is often fractional — at 24 fps and
/// 48 kHz it is exactly <b>12.5</b>, so every odd slot boundary is a midpoint
/// tie, and <b>the rounding rule is part of the wire format</b>. Swift's
/// <c>.rounded()</c> is half-away-from-zero; .NET's default
/// <c>Math.Round(double)</c> is banker's rounding and would place every other
/// boundary one sample early. Hence the explicit
/// <see cref="MidpointRounding.AwayFromZero"/> — it is not a style choice.
/// </para>
/// </remarks>
public static class LtcEncoder
{
    /// <summary>Standalone/test fallback amplitude. The product default lives in
    /// the routing settings and is always passed explicitly on the live path.</summary>
    public const float DefaultAmplitude = 0.8f;

    private const int HalfBitSlotsPerFrame = 160;

    /// <summary>
    /// Float PCM samples (mono, <c>±amplitude</c>) for one LTC frame at
    /// <paramref name="timecode"/>, continuous from <paramref name="startLevel"/>.
    /// Returns the samples and the signal level after the last sample, so
    /// consecutive frames can be modulated continuously.
    /// </summary>
    /// <remarks>Swift traps on a non-positive sample rate
    /// (<c>precondition</c>); throwing here keeps a bad caller failing on both
    /// sides rather than silently producing an empty buffer on one of them.</remarks>
    public static (float[] Samples, bool EndLevel) Samples(
        Timecode timecode,
        double sampleRate,
        float amplitude = DefaultAmplitude,
        bool startLevel = false)
    {
        if (!(sampleRate > 0))
        {
            throw new ArgumentOutOfRangeException(nameof(sampleRate), sampleRate, "sample rate must be positive");
        }

        var frame = LtcFrame.FromTimecode(timecode);
        var halfBitSamples = sampleRate / (HalfBitSlotsPerFrame * (double)timecode.Rate.FramesPerSecond());
        var high = amplitude;
        var low = -amplitude;

        var level = startLevel;
        var samples = new List<float>((int)Round(sampleRate / timecode.Rate.FramesPerSecond()) + 2);
        var slot = 0;

        void EmitSlot()
        {
            var start = Round(slot * halfBitSamples);
            var end = Round((slot + 1) * halfBitSamples);
            for (var count = Math.Max(0, end - start); count > 0; count--)
            {
                samples.Add(level ? high : low);
            }

            slot++;
        }

        foreach (var bit in frame.Bits)
        {
            level = !level;             // bit-boundary transition (always)
            EmitSlot();
            if (bit)
            {
                level = !level;         // mid-bit transition for a 1
            }

            EmitSlot();
        }

        return (samples.ToArray(), level);
    }

    /// <summary>Swift's <c>Double.rounded()</c>: half away from zero, then
    /// truncated to an <c>int</c>.</summary>
    private static int Round(double value) => (int)Math.Round(value, MidpointRounding.AwayFromZero);
}
